nimbus/src/enrollment.rs

```diff
@@ -179,55 +525,117 @@ pub fn reset_enrollment(
     nimbus_id: &Uuid,
     aru: &AvailableRandomizationUnits,
     app_context: &AppContext,
-) -> Result<()> {
+) -> Result<Vec<EnrollmentChangeEvent>> {
     let exp_store = db.get_store(StoreId::Experiments);
     let exp = match exp_store.get::<Experiment>(&writer, experiment_slug)? {
         None => {
             // XXX - do we want specific errors for this kind of thing?
             log::warn!("No such experiment '{}'", experiment_slug);
-            return Ok(());
+            return Ok(vec![]);
         }
         Some(e) => e,
     };
-    let enrollment = evaluate_enrollment(nimbus_id, aru, app_context, &exp)?;
+    let merger = EnrollmentsMerger::new(nimbus_id, aru, app_context);
+    let is_user_participating = get_global_user_participation(db, writer)?;
+    let mut enrollment_events = vec![];
+    let enrollment = merger
+        .evolve_enrollment(
+            is_user_participating,
+            None, // Treat the experiment as we're seeing it for the first time...
+            Some(&exp),
+            None, // ...and we've never been enrolled to it.
+            &mut enrollment_events,
+        )?
+        .ok_or_else(|| {
+            Error::InternalError("reset_enrollment should have resulted in a non-empty enrollment")
+        })?;
```

Comment:
I'm a little concerned about this, but I don't know enough context to comment for sure. Under what circumstances will the consuming app be expected to call this method?

The docstring suggests that it is for undoing an opt-out, but:

* I'm not aware of any product requirement to let to user opt back in to existing experiments after opting out.
  * When this came up in a recent Nimbus Architecture Group meeting, folks seemed pretty happy with the simplification of declaring this impossible, ref [Nov 12 meeting minutes](https://docs.google.com/document/d/12ARuGAa3vXrtS8f7_weg7oUAItREsRr7Vk0fV07JKhc/edit#heading=h.pz2i4nsa829b).
* It can reset much more than just opt-out status, e.g. if you were disqualified for some other reason then this could re-qualify you, or if you were in one branch and your context has changed then you could end up in a different branch.

This method feels like a footgun to me, and this patch is making the footguniness more obvious because of how it lets you make otherwise invalid state transitions. ISTM we should either limit it to just the specifically documented effects (e.g. only resetting states with `reason: OptOut`) or remove it entirely.

@mhammond maybe you can add more context here?

Reply:
I agree this should be removed as a public function. This is also called from `update_enrollments` to handle the case where the *global* opt-out state changes (and that does indeed only do it for the `OptOut` reason.. The meeting docs you linked above don't make it clear whether that decision also applies to the global state or not. If it does, then I agree it can be removed entirely.

Reply:
> The meeting docs you linked above don't make it clear whether that decision also applies to the global state or not

Yes, my understanding is this should also apply to globally opting-out and then globally opting back in - you become eligible for new experiments, but if you were disqualified from an existing experiment then you stay disqualified.